mkosi/__init__.py

```diff
@@ -3398,7 +3453,141 @@ def boot_directory(state: MkosiState, kver: str) -> Path:
+    if magic[:4] == MAGIC_NUMBERS["zstd"]:
+        return "zstd"
+    elif magic == MAGIC_NUMBERS["xz"]:
+        return "xz"
+    return None
+
+
+def list_kernel_modules_in_initrds(state: MkosiState) -> dict[str, Path]:
+    kernel_modules: dict[str, Path] = {}
+    for initrd in state.config.initrds:
+        try:
+            compression = find_compression(initrd)
+            if not compression:
+                raise MkosiException("Unsupported compression")
+            with subprocess.Popen(uncompress_command(compression, initrd), stdout=subprocess.PIPE) as archive:
+                output: str = run(["cpio", "-it"],
+                                  stdin=archive.stdout, stdout=subprocess.PIPE, stderr=subprocess.DEVNULL, check=True, text=True).stdout
+                for s in output.split("\n"):
+                    path = Path(s)
+                    if path.parent == Path("usr/lib/modules"):
+                        kernel_modules[path.name] = initrd
+        except Exception as err:
+            MkosiPrinter.info(f"Couldn't list kernel modules in {initrd}: {err=}")
+    return kernel_modules
+
+
+@dictify
+def read_os_release(root: Path) -> Iterator[tuple[str, str]]:
+    try:
+        f = root.joinpath('etc/os-release').open()
+    except FileNotFoundError:
+        f = root.joinpath('usr/lib/os-release').open()
+
+    for line_number, line in enumerate(f, start=1):
+        if not line.strip() or line.startswith('#'):
+            continue
+        if m := re.match(r'([A-Z][A-Z_0-9]+)=(.*)', line):
+            name, val = m.groups()
+            if val and val[0] in '"\'':
+                val = ast.literal_eval(val)
+            yield name, val
+        else:
+            print(f'Bad line {line_number}: {line}', file=sys.stderr)
+
+
+def update_suffixed(items: dict[str, str], name: str, fallback: str) -> None:
+    value = items.get(name, fallback)
+    if 'mkosi-initrd' in value:
+        return
+    items[name] = value + ' (mkosi-initrd)'
+
+
+def make_initrd_release(root: Path, out: TextIO) -> None:
+    os_release = read_os_release(root)
+
+    # Replacing fields in the original dictionary should maintain the order
+    update_suffixed(os_release, 'NAME', 'Linux')
+    update_suffixed(os_release, 'PRETTY_NAME', 'Linux')
+    os_release['VARIANT'] = 'mkosi-initrd'
+    os_release['VARIANT_ID'] = 'mkosi-initrd'
+
+    for name, value in os_release.items():
+        if value:
+            print(f'{name}={value!r}', file=out)
+
+
+def init_symlink_path(root: Path) -> Path:
+    return root / 'init'
+
+
+def initrd_release_path(root: Path) -> Path:
+    return root / 'etc/initrd-release'
+
+
+def make_init_symlink(root: Path) -> Path:
+    init = init_symlink_path(root)
+    init.unlink(missing_ok=True)
+    init.symlink_to('usr/lib/systemd/systemd')
+    return init
+
+
+def write_initrd_release(root: Path) -> Path:
+    initrd_release = initrd_release_path(root)
+    initrd_release.unlink(missing_ok=True)
```

Comment:
We should not remove existing initrd-release files. Let's only add one if none exists yet